aspnet/security/authorization/secure-data.rst

```diff
@@ -0,0 +1,369 @@
+  :end-before: </snippet_1>
+  :emphasize-lines: 7
+
+Update both ``Delete`` methods to use the authorization filter to verify the user owns the contact.
+
+.. literalinclude:: secure-data/samples/final/Controllers/ContactsController.cs
+  :language: c#
+  :start-after: #region snippet_Delete
+  :end-before: #endregion
+  :dedent: 8
+  :emphasize-lines: 14-19,31-36
+
+.. _update-access-denied--label:
+
+Update the ``AccountController`` to display friendly access denied errors
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Add the ``AccessDenied`` method to the ``AccountController``. This method will be invoked when we call ``ChallengeResult``.
+
+.. literalinclude:: secure-data/samples/final/Controllers/AccountController.cs
+  :language: c#
+  :start-after:  #region snippet_AccessDenied
+  :end-before: #endregion
+  :dedent: 8
+
+Add the *Views/Account/AccessDenied.cshtml* Razor view:
+
+.. literalinclude:: secure-data/samples/final/Views/Account/AccessDenied.cshtml
+  :language: html
+
+Test the ``Edit``, ``Delete``, and ``Create`` methods
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+An easy way to test the changes we made is to launch two different browsers (for example Edge and Internet Explorer). Log into one browser as user ``test@example.com``. In the other browser register a new user (for example ``rick@example.com``) and create a new contact.
+
+Verify ``test@example.com`` can edit and delete the seed data and any contacts created with that account. Verify ``test@example.com`` cannot edit or delete a contact created by the second account.
+
+
+Inject the authorization service into the views
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Currently the UI shows edit and delete links for data the user cannot modify. We'll fix that by applying the authorization filter to the views.
+
+Inject the authorization service in the *Views/_ViewImports.cshtml* file so it will be available to all views:
+
+.. literalinclude:: secure-data/samples/final/Views/_ViewImports.cshtml
+  :language: none
+  :emphasize-lines: 7,8
+
+Update the *Views/Contacts/Index.cshtml* Razor view to show only display the edit and delete links for the users data:
+
+.. literalinclude:: secure-data/samples/final/Views/Contacts/Index.cshtml
+  :language: none
+  :lines: 2
+
+.. literalinclude:: secure-data/samples/final/Views/Contacts/Index.cshtml
+  :language: none
+  :emphasize-lines: 6-10, 12-17
+  :start-after: <snippet_1>
+  :end-before: </snippet_1>
+  :dedent: 16
+
+Test the app with two different browsers and users to verify they cannot edit or delete contacts they didn't create.
+
+.. warning:: Hiding links from users that do not have permission to edit or delete data does not secure that app, it makes the app more user friendly by displaying only valid links. Users can hack the generated URLs to invoke edit and delete operations on data they don't own, it's up to the controller to secure the data.
```

Comment:
"it only makes the application more user friendly". "the controller must repeat the access checks to be secure".